docs/proposals/generalize-modules.md

```diff
@@ -0,0 +1,72 @@
+---
+title: "Generalize Modules Service to make it extensible"
+linkTitle: "Generalize Modules Service to make it extensible"
+weight: 1
+slug: generalize-modules
+---
+
+- Author: @annanay25
+- Reviewers: 
+- Date: April 2020
+- Status: Draft
+
+## Overview
+
+Cortex uses modules to start and operate services with dependencies. Inter-service dependencies are specified in a map and passed to a module manager which ensures that they are started in the right order of dependencies. While this works really well, the implementation is tied in specifically to the Cortex struct and is not flexible for use with other projects like Loki, which also require similar forms of dependency management. 
+ 
+We would like to extend modules in cortex to a generic dependency management framework, that can be used by any project with no ties to cortex. 
+
+## Specific goals
+
+- Framework should allow for reusing cortex modules and allow us to:
+  - Add new modules
+  - Overwrite the implementation of a current module
+  - Manage dependencies
+- Framework should allow for building an application from scratch using the `modules` package, with no dependencies on Cortex. For ex: Remove code from Loki that was copied from `pkg/cortex/cortex.go`.
+
+
+
+## Proposed Design
+
+To make the modules package extensible, we need to abstract away any Cortex specific details from the module manager. The proposed design is to -
+
+- Make a new component `moduleManager`, which is envisioned to be a central manager for all modules of the application. It stores modules & dependencies, and will be housed under a new package `pkg/util/modules`. The following is the interface for interacting with the `moduleManager`:
+```
+// Manager defines the interface for interaction with moduleManager.
+type Manager interface {
+   RegisterModule(name string, deps []string, svc service, options ...Option)
+   AddDependency(fromModule string, toModule string) error
+   ModuleServiceWrapper(name string, modServ services.Service)
+   InitModuleServices(target string) (map[string]services.Service, error)
```

Comment:
It would be nice if this returned a services.Manager since we immediately convert it into one in `Cortex.Run`. As far as I'm aware the only other use for this map is to support the `/services` endpoint. Could the services.Manager be refactored to act as a handler for this endpoint?

Reply:
`InitModuleServices` contract is not clear. What state are returned services in?

> Could the services.Manager be refactored to act as a handler for this endpoint?

`services.Manager` is not a place to put HTTP handlers into. One has many service managers in a running program. Service manager doesn't use a service names, and only needs list of services, not their names.

Reply:
> It would be nice if this returned a services.Manager since we immediately convert it into one in Cortex.Run

I find current proposal (returning map) better – more flexible. Client may be interested in modifying (eg. extra wrapping) some specific services before using them.